File: Python3/CircuitBreaker.py
import requests  # you need to add this to your interpreter through pip
import time
import random


# using a Class to keep state on the CircuitBreaker

#
#          The main thing of this excersize is that the CircuitBreaker can be created in multiple ways.
#          I've used a Class to keep the state of the *state*-machine
#          I've used recursion to call back into the function we just started, with the main difference being:
#             1: Some retry-count was incremented
#             2: The State was changed, so a different codeblock will be executed
#
#          I chose to not use an Enum for Python, just because I got lazy :)

class CircuitBreaker(object):
    def __init__(self, url, time_out, retry_count):
        self.url = url
        self.time_out = time_out
        self.retry_count = retry_count
        self.status = "OPEN"
        self.retries = 0

    """tries to get the actual response from the URL through the requests library"""

    def call(self):
        # try to do HTTP request
        try:
            response = requests.get(self.url)
            rand = random.randrange(0, 2)  # success-rate of 50%
            if rand:  # if rand == 1
                self.status = "SUCCESS"
                return response.json()
            else:
                raise Exception("fake bad response")
        except Exception as e:
            print(e)
            raise e

    """this is the recursive, state-machine function"""

    def start(self):  # state-machine
        print(self.status)

        if self.status == "OPEN":
            try:
                return self.call()
            except Exception as bad_response:
                # do something with bad_response...

                self.status = "HALF-OPEN"
                self.retries += 1
                return self.start()  # lets try again with our status set to HALF-OPEN

        if self.status == "HALF-OPEN":
            time.sleep(self.time_out)
            try:
                return self.call()
            except:
                if self.retries < self.retry_count:  # retry until the retries is exceeded
                    self.retries += 1
                    return self.start()
                else:
                    self.status = "CLOSED"  # set status to CLOSED
                    return self.start()

        if self.status == "CLOSED":  # raise our final error
            raise Exception(f"CircuitBreaker failed for {self.url} : {self.retries} times : timeout {self.time_out}")


# we create a new CB
cb = CircuitBreaker('https://jsonplaceholder.typicode.com/todos', 1, 3)

try:
    response = cb.start()
    print('response =', response)
except Exception as circuit_breaker_exception:
    print(circuit_breaker_exception)

print()
print("total retries:", cb.retries)
print("final status:", cb.status)

File: CSHARP/CircuitBreaker/CircuitBreaker/CircuitBreaker.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Flurl.Http;

namespace CircuitBreaker
{
    public class CircuitBreaker
    {
        public string Url { get; set; }
        public int RetryCount { get; set; }
        public int TimeOut { get; set; }
        public int Retries { get; set; }
        public Status Status { get; set; }


        /*
         Call() actually calls the given URL and returns the result if succesful.
         Also sets the Status to Status.SUCCESS
         */
        public async Task<Todo> Call()
        {
            var rand = new Random().Next(1, 11); // get a random number between 1 and 10
            Console.WriteLine("random int: " + rand);
            if (rand > 5) // success-rate of 50%, if number > 5
            {
                Status = Status.SUCCESS;
                // Url is a string, with Flurl.Http we can just call GetJsonAsync and do a Get request on the URL.
                return await Url.GetJsonAsync<Todo>().ConfigureAwait(false);
            }
            // if rand <= 5 throw a fake bad-response error (e.g.)
            throw new Exception("fake bad response");
        }


        // Recursive, retry-logic, state-machine function
        public async Task<Todo> Start()
        {  
            // let's keep track of our function's status
            Console.WriteLine();
            Console.WriteLine("status: " + Status.ToString());
            Console.WriteLine("retries: " + Retries.ToString());

            // if Start() is called with the CircuitBreaker status being OPEN
            if (Status == Status.OPEN)
            {
                try
                {
                    // lets do our initial call
                    return await Call().ConfigureAwait(false);
                }
                catch (Exception badResponse)
                {
                    // do something with bad_response...
                    Console.WriteLine(badResponse.Message);
                    // set status to HALFOPEN because our initial request failed
                    Status = Status.HALFOPEN;
                    Retries++;
                    return await Start().ConfigureAwait(false); // retry with our Status being HALFOPEN
                }
            }

            // if Start() is called with the CircuitBreaker status being HALFOPEN
            if (Status == Status.HALFOPEN)
            {
                Console.WriteLine("TIMEOUT...");
                Thread.Sleep(TimeOut);
                try
                {
                    // try a call again
                    return await Call().ConfigureAwait(false);
                }
                catch
                {
                    if (Retries < RetryCount) // did we retry too many times already?
                    {
                        Retries++;
                        return await Start().ConfigureAwait(false); // Start this function again with the same status HALFOPEN
                    }

                    // looks like we retried too many times, time to set status to CLOSED
                    Status = Status.CLOSED;
                    return await Start().ConfigureAwait(false);
                }
            }

            // if Start() is called with the CircuitBreaker status being CLOSED
            if (Status == Status.CLOSED)
            {
                // we throw our final error message
                throw new Exception($"CircuitBreaker failed for { Url} : { Retries} " +
                    $"times: TimeOut { TimeOut}");
            }

            return await Start().ConfigureAwait(false); // this should never be reached, but acts as a safety-net
        }
    }
}
